dvc/output/local.py

```diff
@@ -8,70 +8,70 @@
 from dvc.istextfile import istextfile
 from dvc.exceptions import DvcException
 from dvc.remote.local import RemoteLOCAL
-from dvc.output.base import OutputBase
+
+from dvc.pkg import Pkg
+
+from . import Output
 
 
 logger = logging.getLogger(__name__)
 
 
-class OutputLOCAL(OutputBase):
+class OutputLOCAL(Output):
     REMOTE = RemoteLOCAL
 
-    def __init__(
-        self,
-        stage,
-        path,
-        info=None,
-        remote=None,
-        cache=True,
-        metric=False,
-        persist=False,
-        tags=None,
-    ):
+    PARAM_PKG = "pkg"
+
+    def __init__(self, stage, info, remote=None, dependency=False):
+        pkg_info = info.pop(self.PARAM_PKG, None)
+
         super(OutputLOCAL, self).__init__(
-            stage,
-            path,
-            info,
-            remote=remote,
-            cache=cache,
-            metric=metric,
-            persist=persist,
-            tags=tags,
+            stage, info, remote=remote, dependency=dependency
         )
         if remote:
             p = os.path.join(
                 remote.prefix, urlparse(self.url).path.lstrip("/")
             )
         else:
-            p = path
+            p = self.url
+
+        if not pkg_info:
+            self.pkg = None
+            if not os.path.isabs(p):
+                p = self.remote.to_ospath(p)
+                p = os.path.join(stage.wdir, p)
+            p = os.path.abspath(os.path.normpath(p))
+
+            self.path_info = PathLOCAL(url=self.url, path=p)
+            return
+
+        assert self.dependency
 
-        if not os.path.isabs(p):
-            p = self.remote.to_ospath(p)
-            p = os.path.join(stage.wdir, p)
-        p = os.path.abspath(os.path.normpath(p))
+        self.pkg = Pkg.create(stage.repo, pkg_info)
+        if not self.pkg.installed:
+            self.pkg.install()
 
-        self.path_info = PathLOCAL(url=self.url, path=p)
+        out_path = os.path.join(
+            self.pkg.repo.root_dir, urlparse(self.url).path.lstrip("/")
+        )
+
+        out, = self.pkg.repo.find_outs_by_path(out_path)
+        self.path_info = PathLOCAL(url=self.url, path=out.cache_path)
+        if not os.path.exists(out.cache_path):
```

Comment:
One thing that bothers me is an additional copy of the data files. In some cases, ppl will use a package for huge datasets like 140Gb imagenet. I don't think it is a good idea of creating two copies localy.

In the previous realization, all the packages use the project cache, not own. This method has its own disadvantages - no local duplication but still global duplication (the project cache\s3 will get another copy of the dataset).

It feels like we are lacking some more holistic approach. Any thoughts?

Reply:
A nice thing about the current approach is that you at least have your cache separate from pkg cache, so it is easy to just `dvc pkg uninstall mypkg` which will delete .dvc/pkg/mypkg along with all of its cache.

But yeah, the additional copy is not ideal, it is just what we have in the initial pkg implementation. Optimizations will come, especially since they are pretty straightforward(e.g. just setting external cache dir for pkgs to .dvc/cache of the main project).

Reply:
After IRL discussion, decided to go with using project's cache for pkgs.